### PR TITLE
Feature/FFU-1/enter leave space

### PR DESCRIPTION
**Pull request by RamsesDijkstra**
should be it for entering and leaving a space

### CODE REVIEW

File: universe/api/dto/dto.go [hunk 1]
```diff
@@ -170,6 +174,13 @@ type Space struct {
 	} `json:"metadata,omitempty"`
 }
 
+type SpaceAttribute struct {
+	PluginID      string `json:"pluginId"`
+	AttributeName string `json:"attribute_name"`
+	SpaceID       string `json:"stringId"`
+	Value         any
+}
+
```
Comment (cnaize) [inline universe/api/dto/dto.go]: do we need it?

File: universe/node/api_space_attributes.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	if err != nil {
+		err := errors.WithMessage(err, "Node: apiGetSpaceAttributes: failed to parse space id")
+		api.AbortRequest(c, http.StatusBadRequest, "invalid_space_id", err, n.log)
+		return
+	}
+
+	inQuery := struct {
+		PluginID string `form:"plugin_id" binding:"required"`
+		Name     string `form:"attribute_name" binding:"required"`
+	}{}
```
Comment (cnaize) [inline universe/node/api_space_attributes.go]: please, move it to first line off the method
it's good practice to put all declarations first

File: universe/node/api_space_attributes.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	if !ok {
+		err := errors.Errorf("Node: apiGetSpaceAttributes: space not found: %s", spaceID)
+		api.AbortRequest(c, http.StatusNotFound, "space_not_found", err, n.log)
+		return
+	}
+
+	attributeID := entry.NewAttributeID(pluginID, inQuery.Name)
+	out, ok := space.GetSpaceAttributeValue(attributeID)
+	if !ok {
+		err := errors.Errorf("Node: apiGetSpaceSubAttribute: SubAttributeKey not found: %s", attributeID)
```
Comment (cnaize) [inline universe/node/api_space_attributes.go]: wrong error message

File: universe/node/api_space_attributes.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		err := errors.WithMessage(err, "Node: apiGetSpaceSubAttributes: failed to parse space id")
+		api.AbortRequest(c, http.StatusBadRequest, "invalid_space_id", err, n.log)
+		return
+	}
+
+	inQuery := struct {
+		PluginID        string `form:"plugin_id" binding:"required"`
+		Name            string `form:"attribute_name" binding:"required"`
+		SubAttributeKey string `form:"sub_attribute_key" binding:"required"`
+	}{}
```
Comment (cnaize) [inline universe/node/api_space_attributes.go]: the same as above

File: universe/node/api_space_attributes.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	if !ok {
+		err := errors.Errorf("Node: apiGetSpaceSubAttribute: space not found: %s", spaceID)
+		api.AbortRequest(c, http.StatusNotFound, "space_not_found", err, n.log)
+		return
+	}
+
+	attributeID := entry.NewAttributeID(pluginID, inQuery.Name)
+	attributeValue, ok := space.GetSpaceAttributeValue(attributeID)
+	if !ok {
+		err := errors.Errorf("Node: apiGetSpaceSubAttribute: AttributeValue not found: %s", attributeID)
```
Comment (cnaize) [inline universe/node/api_space_attributes.go]: wrong error message

File: universe/node/api_space_attributes.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		api.AbortRequest(c, http.StatusNotFound, "sub_attribute_value_not_found", err, n.log)
+		return
+	}
+
+	if attributeValue == nil {
+		err := errors.Errorf("Node: apiGetSpaceSubAttribute: AttributeValue is nil")
```
Comment (cnaize) [inline universe/node/api_space_attributes.go]: please, use lower case for `AttributeValue`

File: universe/node/api_space_attributes.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+		err := errors.Errorf("Node: apiGetSpaceAttributes: space not found: %s", spaceID)
+		api.AbortRequest(c, http.StatusNotFound, "space_not_found", err, n.log)
+		return
+	}
+
+	attributeID := entry.NewAttributeID(pluginID, inQuery.Name)
+	out, ok := space.GetSpaceAttributeValue(attributeID)
+	if !ok {
+		err := errors.Errorf("Node: apiGetSpaceSubAttribute: space_attribute_value not found: %s", attributeID)
+		api.AbortRequest(c, http.StatusNotFound, "sub_attribute_not_found", err, n.log)
```
Comment (cnaize) [inline universe/node/api_space_attributes.go]: just `attriubute` not `sub` :)

**Review by cnaize — CHANGES_REQUESTED** [review]
please, use, for example, `sub attribute value` for messages instead of `SubAttributeKey` and `attribute_value`, ...

**Review by cnaize — APPROVED** [review]

File: universe/api/dto/dto.go [hunk 8]
```diff
@@ -170,6 +174,13 @@ type Space struct {
 	} `json:"metadata,omitempty"`
 }
 
+type SpaceAttribute struct {
+	PluginID      string `json:"pluginId"`
+	AttributeName string `json:"attribute_name"`
+	SpaceID       string `json:"stringId"`
```
Comment (boyan-d) [inline universe/api/dto/dto.go]: shouldn't the json tag be `"spaceId"` not `"stringId"`?

Comment (RamsesDijkstra) [reply universe/api/dto/dto.go]: this struct was removed :) but good catch :D

File: universe/node/api_space_attributes.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
+package node
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/google/uuid"
+	"github.com/momentum-xyz/ubercontroller/types/entry"
+	"github.com/momentum-xyz/ubercontroller/universe/api"
+	"github.com/pkg/errors"
+	"net/http"
```
Comment (boyan-d) [inline universe/node/api_space_attributes.go:9]: group imports - 1st: standard lib, 2nd other imports, 3rd internal ones

**Pull request merged**